### DISCUSSION TITLE
gpsgraphic filter is not showing anything, and background image is distorted

### DISCUSSION THREAD

**Discussion by shadowfax65** [mltframework/mlt#1109] [General]
I am using Kdenlive and am trying to get the **gpsgraphic** filter to work. Kdenlive version 24.08.3 with MLT Framework version 7.27.0
My example .gpx file is a download from alltrails.com of a hike I did this year. There is no output whatsover. The only thing that sort of works is adding a background image, but it is distorted and only displayed in the bottom left-hand corner of the rectangle for the filter.

**Comment by ddennedy** [comment]
The developer of this filter @dany123  primarily uses Shotcut and may not be able to debug it through Kdenlive. Not all GPX are the same as the X in XML means extensible, similar to how MLT XML differs between projects that use it. Can you provide a sample of the AllTrails GPX?

**Comment by shadowfax65** [comment]
Thanks for your quick reply. Much appreciated.

[example_gpx.zip](https://github.com/user-attachments/files/17740762/example_gpx.zip)

**Comment by dany123** [comment]
As Dan says I am only using Shotcut so I won't be able to help with Kdenlive specific stuff.

I can load and play your .gpx in Shotcut with no problem so there's no bug with this file in regards to the backend. The map should show up with no other changes other than loading the file. Shotcut's UI does do some helpful stuff but it should work even without them.

Here's the filter's properties in the saved .mlt. The only thing I changed was the filter's rectangle to have the map with the correct aspect ratio.

```
<filter id="filter0" out="00:19:26.967">
      <property name="resource">C:/Users/Daniel/Desktop/browser_downloads/Afternoon hike at Bondi Beach to Coogee Beach Walk.gpx</property>
      <property name="time_offset">1713748602</property>
      <property name="smoothing_value">5</property>
      <property name="speed_multiplier">1</property>
      <property name="graph_data_source">0</property>
      <property name="graph_type">0</property>
      <property name="trim_start_p">0</property>
      <property name="trim_end_p">100</property>
      <property name="crop_mode_h">0</property>
      <property name="crop_left_p">0</property>
      <property name="crop_right_p">100</property>
      <property name="crop_mode_v">0</property>
      <property name="crop_bot_p">0</property>
      <property name="crop_top_p">100</property>
      <property name="color_style">1</property>
      <property name="color.1">#00aaff</property>
      <property name="color.2">#ffffff</property>
      <property name="color.3">#ffff00</property>
      <property name="color.4">#ff8c00</property>
      <property name="color.5">#ff0000</property>
      <property name="show_now_dot">1</property>
      <property name="now_dot_color">#00ffffff</property>
      <property name="show_now_text">0</property>
      <property name="angle">0</property>
      <property name="thickness">5</property>
      <property name="rect">160 16 504 964 1</property>
      <property name="show_grid">0</property>
      <property name="legend_unit"></property>
      <property name="draw_individual_dots">0</property>
      <property name="map_coords_hint">-33.904377, 151.268395</property>
      <property name="bg_img_path">!</property>
      <property name="bg_scale_w">1</property>
      <property name="bg_opacity">1</property>
      <property name="mlt_service">gpsgraphic</property>
      <property name="majoroffset_sign">1</property>
      <property name="gps_start_text">2024-04-22 01:16:42</property>
      <property name="map_original_aspect_ratio">0.522711</property>
      <property name="video_start_text">1970-01-01 00:00:00</property>
      <property name="auto_gps_offset_start">1713748602</property>
      <property name="auto_gps_offset_now">1713748602</property>
    </filter>
```

Here's how I see it after load (ignore the GPS offset's huge value, it's relative to the Color background's "default" date):
![image](https://github.com/user-attachments/assets/05eb45a4-8b65-4e88-a2be-46c91a5e2689)

**Comment by dany123** [comment]
As for the background, there's a little helpful hint in the shotcut's UI as the center of the background img has to match the center of the GPX track so there's a button that shows this info, then you need to manually scale to match it on the image. But other than this, it should properly load without any issue as long as the frontend sends the right path to it.
! GPS file centre is: -33.904377, 151.268395

![image](https://github.com/user-attachments/assets/608d1276-11dd-42ed-bf1f-f7c57559cddb)

**Comment by shadowfax65** [comment]
That was quick! Thank you very much! I will show this to the Kdenlive dev so that he can look into it more.

**Comment by dany123** [comment]
I've had a look in kdenlive and as far as I can tell the UI was automatically generated based on the available fields in the filter's description and some important fields - mainly the one that should point to the GPS file is wrong/not usable for me (but also a field that controls the time offset should be present and editable).

I've manually edited the project file with notepad to add this line to the proper filter inside:
`      <property name="resource">C:/Users/Daniel/Desktop/browser_downloads/Afternoon hike at Bondi Beach to Coogee Beach Walk.gpx</property>`

And something showed up but it was only 1px in size. But I could then drag it with the mouse and the map (gps track) showed up.

So the main functionality seems to be there, the issue is there's a lot of fields either missing, with bad defaults or not formatted/grouped nicely. This particular filter is quite complex so the automatic generation is just too basic for it to work. 
If a kdenlive contributor wants to fix this I recommend to start by looking at how the filter looks in Shotcut and maybe ignore some of the more advanced fields like cropping and colors.

**Comment by shadowfax65** [comment]
Thanks for looking into this even deeper, @dany123.

I am trying to fix the issue you described by creating the XML file Kdenlive uses for the UI.
[gpsgraphic_xml.zip](https://github.com/user-attachments/files/17749835/gpsgraphic_xml.zip)

I used the documentation available at https://www.mltframework.org/plugins/FilterGpsgraphic/. The only fields I did not include were the ones without a type specification but they appeared to be more informative. If you let me know how they are used I can include them.

With the XML file, Kdenlive displays a proper UI, lets the user select the .gpx and a background file via the normal file selection UI. The only thing that might be an issue is the rectangle which I implemented as `animatedrect` and which perhaps should be `geometry`.

**Comment by dany123** [comment]
This looks way better and is functional. Just a few quick notes:

-make sure to set the defaults for trim end, crop right, crop bottom at 100 (so the first time the gpx file is loaded it will actually show a map)
-for the crop left/right/top/bottom I think the sliedr would be more useable with a smaller range (I'd say -200 to 200 would be enough for 99% of cases). But it's your choice, the values are valid anyway.
-the Crop top field is wrongly named Crop right
-I'd say the "show now dot" should be enabled by default (to show the current location)

As for the read-only fields you mention the only really important one would be: map_coords_hint. Without it it becomes a huge pain to put a correctly centered background image.
To properly show it you must wait for it to be populated after loading a gpx file, in Shotcut there's a button that when pressed it will show a text inside a textbox with the string: "GPS file centre is: -33.904377, 151.268395", you could probably do the same thing with a small questionmark button or something like that near the background buttons.

**Comment by dany123** [comment]
@shadowfax65 did you push this into kdenlive yet? 
I think we can close this issue as it's not a bug in mlt.

**Comment by shadowfax65** [comment]
Yes! Thank you for your support.

**Comment by shadowfax65** [comment]
@dany123, sorry to bother you again but I am wondering whether the trim parameters are animated. That would allow to actually draw the line of the path.

**Comment by dany123** [comment]
Can you explain what exactly are you trying to do? From your description the filter should already be able do it without having to use trim.

If you want to quickly draw the path you can increase the speed to something like 100 and it will be drawn in a few seconds, you can further tweak this to hide the now_dot so you only see the path and you can set the first color (while using the two color "past/future" style) to something fully transparent so there will be a "continuous" drawing of the path.

**Comment by shadowfax65** [comment]
I use the filter on an 8-second clip (satellite image of the route). The trip took several hours to do. The speed multiplier is capped at 30 but I cranked that up to 1000 or even higher.
The entire path is drawn on the first frame, the Now Dot doesn't move at all, regardless of the speed multiplier (I tested it with a multiplier of 10,000 based on a 4-hour trip and 8-sec clip duration - something must change, right?). Perhaps something is wrong with how the filter is implemented in Kdenlive.

Assuming that the path is always drawn completely from the first frame, the `trim_end` parameter could be animated to start at 0% and ending at 100%, effectively drawing the path over the duration of the clip the filter is applied to. Which, of course, is not needed if the path should be drawn according to the position/timecode in the `gpx` file.

**Comment by dany123** [comment]
Here's the result of an ~8h gps track at 3600x (so 1 hour = 1 second) using the 2 things I mentioned before: hide the now_dot, and set one of the colors (the 2nd color in my case) to full transparent (so alpha=0 in the color picker dialog):

https://github.com/user-attachments/assets/752c95f7-e075-4751-9f97-ab2e1ebd0996

If nothing changes on your filter even at 10.000x I have to assume that your "GPS offset" is not correct (it's a filter property that has to be set to some value depending on the file created date). It's probably set in the future so the "track" has already been finished, change it to something negative and test it out. 

There's a helper filter property that you can set so the graph starts exactly at the begining of your image: open your project file in notepad and look for the property named "auto_gps_offset_start":
`      <property name="auto_gps_offset_start">1583396338</property>`
then copy the numeric value from there into the property named "time_offset":
`      <property name="time_offset">1583396338</property>`

Then reopen your project in kdenlive and hit play, there should be some movement immediately.
(in Shotcut I've made the UI to load this automatically one time at file load and there's also a button that syncs this to the start)

**Comment by dany123** [comment]
Regarding animating trim_end, that could also work but I have no idea how kdenlive does keyframing/animation so I can't help there.

**Comment by shadowfax65** [comment]
Sorry for the late reply. I got it to work with a **Speed Multiplier** of about 1,800. The **Now Dot** is moving along the track nicely. I also fiddled with the **Graph Color Style** and got the path to draw. However, if I set **Graph Color Style** to, say, _Color by Speed_ or _by Altitude_, the path stays fixed. This is where an animated **Trim End** parameter would fit nicely ;-)
The strange thing is that if I make **Trim End** animated, the whole effect/filter doesn't work anymore in Kdenlive. I will discuss this with the dev team.

**Reply by dany123** [reply]
> However, if I set **Graph Color Style** to, say, _Color by Speed_ or _by Altitude_, the path stays fixed

Correct,  these use a different coloring logic. it's on my todo list to update this to also follow past/future like the 2 color mode.

 >  The strange thing is that if I make Trim End animated, the whole effect/filter doesn't work anymore in Kdenlive. I will discuss this with the dev team.

Maybe it's defaulting to 100 so the entire path is hidden?
